gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java

```diff
@@ -190,7 +192,9 @@ void handleAttempt(Throwable throwable, ResponseT response) {
           setAttemptResult(throwable, response, true);
           // a new attempt will be (must be) scheduled by an external executor
         } else if (throwable != null) {
-          if (retryAlgorithm.getResultAlgorithm().shouldRetry(throwable, response)) {
+          if (retryAlgorithm
+              .getResultAlgorithmWithContext()
```

Comment:
I see, the comment with the new returned type justifies the creation of the new method in this case. I was hopping maybe it was not, but looks like changing the returned type, even if the returned type is strictly a subclass is a breaking change: (https://stackoverflow.com/questions/47476509/can-i-change-a-return-type-to-be-a-strict-subtype-and-retain-binary-compatibilit).

Then maybe only one nitpicking comment: WDYT about reanaming this method (and the other similar methods, if any) from `getResultAlgorithmWithContext()` to `getContextAwareResultAlgorithm()`? Otherwise the method's name sounds like it also returns context, while it does not.